java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BUILD.bazel

```diff
@@ -1,58 +1,46 @@
 load("@rules_java//java:defs.bzl", "java_binary", "java_library")
 load("//java/private:artifact.bzl", "artifact")
 
-# We want to be able to compile the test runner, but we don't want to
-# expose any of its dependencies, since we expect whoever is running
-# tests to add them themselves, and we don't want to pollute the classpath.
-# As such, we do this rather weird thing to make sure we have the runner
-# but none of its dependencies.
-java_import(
-    name = "junit5",
-    jars = [
-        ":junit5-copy-file",
-    ],
-    visibility = [
-        "//visibility:public",
-    ],
-    deps = [
-        ":allow",
-    ],
-)
-
 java_library(
-    name = "allow",
-    srcs = ["allow.java"],
-    # At some point the SecurityManager class is going to go away
+    name = "junit5",
+    srcs = glob(
+        ["*.java"],
+        exclude = [
+            "Java17SystemExitToggle.java",
+            "SystemExitToggle.java",
+        ],
+    ),
     javacopts = [
+        # Target Java 8, no matter which JDK we're on so that we
+        # can be sure that `SecurityManager` is present, even if
+        # someone rebuilds on a JDK that doesn't have it.
         "--release",
-        "11",
+        "8",
     ],
+    visibility = ["//visibility:public"],
     deps = [
-        # There must never be dependencies
+        ":maven-deps-compile-only",
+        ":system-exit-toggle",
     ],
 )
 
-genrule(
-    name = "junit5-copy-file",
-    srcs = [":junit5-compile"],
-    outs = ["junit5-runner.jar"],
-    cmd = "cp $< $@",
+# Do not build with --release 8 to work around https://bugs.openjdk.org/browse/JDK-8214165.
+# Files in srcs must not reference SecurityManager as it will go away soon.
+java_library(
+    name = "system-exit-toggle",
+    srcs = [
+        "Java17SystemExitToggle.java",
+        "SystemExitToggle.java",
+    ],
 )
 
+# We want to be able to compile the test runner, but we don't want to
+# expose any of its dependencies, since we expect whoever is running
+# tests to add them themselves, and we don't want to pollute the classpath.
 java_library(
-    name = "junit5-compile",
-    srcs = glob(["*.java"]),
-    javacopts = [
-        # Target Java 11, no matter which JDK we're on so that we
-        # can be sure that `SecurityManager` is present, even if
-        # someone rebuilds on a JDK that doesn't have it.
-        "--release",
-        "11",
-    ],
-    deps = [
-        # The only dependencies here are those required to run
-        # a junit5 test. We try not to pollute the classpath, so
-        # be very careful when adding new deps here.
+    name = "maven-deps-compile-only",
+    neverlink = True,
```

Comment:
Won't this still bleed the transitive deps on to the classpath? That's the thing we're trying to avoid.

Reply:
This adds the transitive deps to the compilation classpath, but if consumers depend on them strict deps checks will catch that and fail the build. `neverlink` ensures that the transitive deps will not be on the runtime classpath.

The only difference I can see is that this could allow tests to build if they accidentally depend on the transitive deps of this target *and* a non-default value of `--strict_java_deps` is used. But even in that case, running the test would still fail since the transitive deps are not on the runtime classpath.

Let me know if you still prefer the old approach and I can revert back to it.